File: leetcode/234.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: ListNode) -> bool:

        
        def solve(head):
            if head:
                ret = solve(head.next)
                ret.append(head.val)
                return ret
            else:
                return []


        ret = solve(head)
        n = len(ret)
        
        for i in range(n):
            if ret[i] != ret[n-i-1]:
                return False
        return True

File: leetcode/234.cpp
#include <bits/stdc++.h>

using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode()
        : val(0)
        , next(nullptr)
    {
    }
    ListNode(int x)
        : val(x)
        , next(nullptr)
    {
    }
    ListNode(int x, ListNode* next)
        : val(x)
        , next(next)
    {
    }
};

class Solution {
public:
    bool isPalindrome(ListNode* head)
    {
        stack<int> stk;

        ListNode* cur = head;

        while (cur) {
            stk.push(cur->val);
            cur = cur->next;
        }

        cur = head;

        while (!stk.empty() && cur) {
            if (stk.top() != cur->val)
                return false;
            stk.pop();
            cur = cur->next;
        }

        return true;
    }
};
